paddle/fluid/platform/nccl_helper.h

```diff
@@ -160,6 +160,147 @@ struct NCCLContextMap {
+}
+inline std::string GetHierarchicalInterNCCLVarName() {
+  return string::Sprintf("Hierarchical_inter_%s", NCCL_ID_VARNAME);
+}
+
+class MultiNCCLContextMap {
+ public:
+  MultiNCCLContextMap() {}
+  virtual ~MultiNCCLContextMap() {
+    for (size_t i = 0; i < flat_ctxs_.size(); i++) {
+      delete flat_ctxs_[i];
+    }
+    flat_ctxs_.clear();
+
+    for (size_t i = 0; i < h_inter_ctxs_.size(); i++) {
+      delete h_inter_ctxs_[i];
+    }
+    h_inter_ctxs_.clear();
+
+    for (size_t i = 0; i < h_exter_ctxs_.size(); i++) {
+      delete h_exter_ctxs_[i];
+    }
+    h_exter_ctxs_.clear();
+  }
+
+  NCCLContextMap *DefaultFlatCtx() const {
+    if (flat_ctxs_.size() == 0) {
+      return nullptr;
+    }
+
+    return flat_ctxs_[0];
+  }
+
+  std::vector<NCCLContextMap *> *GetFlatCtxs() { return &flat_ctxs_; }
+
+  NCCLContextMap *GetFlatCtx(size_t run_order) const {
+    return flat_ctxs_[run_order % flat_ctxs_.size()];
+  }
+
+  NCCLContextMap *GetRunEnvNCCLCtx(size_t run_order,
+                                   bool use_hierarchical_allreduce) const {
+    if (!use_hierarchical_allreduce) {
+      return GetFlatCtx(run_order);
+    }
+
+    return GetHierarchicalInterCtx(run_order);
+  }
+
+  void InitFlatCtxs(const std::vector<platform::Place> &places,
+                    const std::vector<ncclUniqueId *> &nccl_ids,
+                    size_t trainers_num, size_t trainer_id) {
+    if (nccl_ids.size() == 0) {
+      auto ptr = new platform::NCCLContextMap(places);
+      VLOG(1) << "init local trainer";
+      flat_ctxs_.push_back(ptr);
+      return;
+    }
+
+    for (size_t i = 0; i < nccl_ids.size(); i++) {
+      auto ptr = new platform::NCCLContextMap(places, nccl_ids[i], trainers_num,
+                                              trainer_id);
+      VLOG(1) << "init trainer_id:" << trainer_id << ", comm no:" << i;
+      flat_ctxs_.push_back(ptr);
+    }
+  }
+
+  void InitHierarchicalCtxs(const std::vector<platform::Place> &places,
+                            ncclUniqueId *inter_nccl_id,
+                            const std::vector<ncclUniqueId *> &exter_nccl_id,
+                            size_t trainers_num, size_t trainer_id,
+                            size_t inter_trainers_num,
+                            size_t exter_trainers_num) {
+    PADDLE_ENFORCE(trainers_num == inter_trainers_num * exter_trainers_num,
+                   "trainers_num:%llu != inter_trainers_num:%llu * "
+                   "exter_trainers_num:%llu",
+                   trainers_num, inter_trainers_num, exter_trainers_num);
+
+    PADDLE_ENFORCE(inter_trainers_num > 1, "inter_trainers_num:%llu must > 1",
+                   inter_trainers_num);
+
+    int inter_trainer_id = trainer_id % inter_trainers_num;
+    VLOG(1) << "init inter_trainer_id:" << inter_trainer_id;
+    auto local = new NCCLContextMap(places, inter_nccl_id, inter_trainers_num,
+                                    inter_trainer_id);
+
+    h_inter_ctxs_.push_back(local);
```

Comment:
I prefer to use std::unique_ptr here.

Reply:
Free them in destructor function and the vector destructs itself.